saleor/cart/utils.py

```diff
@@ -65,20 +77,34 @@ def check_product_availability_and_warn(request, cart):
         remove_unavailable_variants(cart)
 
 
-def find_and_assign_anonymous_cart(request, queryset=Cart.objects.all()):
+def find_and_assign_anonymous_cart(queryset=Cart.objects.all()):
     """Assign cart from cookie to request user
     :type request: django.http.HttpRequest
     """
-    token = request.get_signed_cookie(Cart.COOKIE_NAME, default=None)
-    if not token:
-        return
-    cart = get_anonymous_cart_from_token(token=token, cart_queryset=queryset)
-    if cart is None:
-        return
-    cart.change_user(request.user)
-    carts_to_close = Cart.objects.open().filter(user=request.user)
-    carts_to_close = carts_to_close.exclude(token=token)
-    carts_to_close.update(status=Cart.CANCELED, last_status_change=now())
+    def get_cart(view):
+        @wraps(view)
+        def func(request, *args, **kwargs):
+            response = view(request, *args, **kwargs)
+            token = request.get_signed_cookie(Cart.COOKIE_NAME, default=None)
+            if not token or token is not None and not token_is_valid(token):
```

Comment:
If you are authenticated and don't have a cookie token, then database cart is taken into account